contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,174 @@
+
+1. Reconstruct Job by `OwnerReference` of `Pod` and `SchedulingSpec`; the objects (`Pods` and `SchedulingSpec`) with same `OwerReference` are considered belonging to the same job
+2. If `SchedulingSpec` is not created, the job will not be scheduled
+3. In `predicate` phase, filter nodes for each job; if no `NodeSelector`, all nodes are candidates
+4. In `allocate` phase, `kar-schedular` will allocate resource to job; and bind `Pod` to host until job ready (`minAvailable` <= `allocated Pods`)
+5. If can not allocate enough resources to the job, the resource can not allocate to other job until next scheduling cycle
+
+That may make resources (less than job's resource request) idle for a while, e.g. a huge job. One of solution is to backfill other smaller jobs to improve the resource utilization.
+
+### kar-controllers (controllers component of kube-arbitrator)
+
+Similar to `kube-controller-manager`, `kar-controller` is a collection of controllers, including `QueueJobController` and `QueueController` (in future).  `QueueJobController` manages the lifecycle of `QueueJob` as following:
+
+* **Pod/QueueJob Creations**: The `QueueJobController` will create pods up to `.spec.SchedSpec.minAvailable` ordered by priority in `spec.TaskSpecs.Template`, and waiting for `kar-scheduler` to schedule pods in batch. `QueueJobController` will handle `QueueJob` by priority: creating `.spec.SchedSpec.minAvailable` pods for QueueJob firstly, and then creating others pods (exceed `.spec.SchedSpec.minAvailable`) in round robin. When exceed `ResourceQuota`, that makes sure only the `QueueJob` with lowest priority is impact.  After creating, `QueueJob`'s (`.spec.SchedSpec.minAvailable`) maybe scale up; if `Quota` exceeded, the scheduler will not schedule the `QueueJob` for the new pods until all pods are created.
+
+* **Pod Failures/Deletion/Unschedulable**: When a pod was failed, deleted or unschedulable, `QueueJobController` manages `QueueJob`'s lifecycle according to its status:
+
+  1. `Number of Running pods` < `.spec.SchedSpec.minAvailable` : If the number of running pods less then `.spec.SchedSpec.minAvailable`, the `QueueJobController` will kill the whole QueueJob and recreate pods for rescheduling in batch.
+
+  2. `Number of Running pods` >= `.spec.SchedSpec.minAvailable`: If there are still enough running pods for the `QueueJob`, the `QueueJobController` will only recreate failed pods, and wait for scheduler to bind the pending pods.
+
+* **QueueJob Update/Delete**: There's not rolling update for `QueueJob` ; if pod template was updated, the whole `QueueJob` is re-created. If `QueueJob` was deleted, all its pods and `SchedulingSpec` will be cleanup.
+
+
+![queuejob](images/gang-scheduling-queuejob.png)
+
+
+### Customized controller
```

Comment:
We also need to cover admission in this document. I think we ideally want to admit either the whole gang or none of the members of a gang. Doing so looks challenging to me with the existing admission controllers. Do you have any ideas on addressing this issue?

Reply:
Transactional multiobject admission is not possible.

Reply:
Would this mean that we have to necessarily have a wrapper (maybe QueueJob) that will allow for admission of a gang of pods in one shot?

Reply:
by the time scheduling happens, the pod objects have already been admitted, which means object counts, resource quotas, and pod spec restrictions have already been enforced... if the gang scheduler is just matching up pods to nodes, it doesn't seem like admission would cause a lot of problems

> Would this mean that we have to necessarily have a wrapper (maybe QueueJob) that will allow for admission of a gang of pods in one shot?

no, you can't end-run around individual pod admission via another object

Reply:
> I think we ideally want to admit either the whole gang or none of the members of a gang. 

It's better to delegate this to controllers or [Applications](https://github.com/kubernetes-sigs/application); it's hard for admit to collect "dynamic info", e.g. for `ResourceQuota`, it's hard for admit to decide which pod should be create to meet 'Gang/PodSchedulingGroup' as much as possible. In google doc, the QueueJob will stop creating new Pod if rejected (there're several corner case to cover)

Reply:
> In google doc, the QueueJob will stop creating new Pod if rejected (there're several corner case to cover)

The problem is that the user experience would be poor. A few pods of a gang are created, but the rest fail admission and are not created. Those pod created will not be scheduled, but will consume the user's quota. I guess the user has to delete them manually. If we cannot have transactional admission, at lease we should have a controller clean up such pods.

Reply:
> A few pods of a gang are created, but the rest fail admission and are not created. 

For batch or run-to-complete workload, it's better to have some pending jobs there; so k8s can start the pending jobs immediately
when some running jobs finished, instead of waiting for client to submit new jobs; hmm...., a kind of pipeline. The difference is that
the "controller" has to make sure not introduce deadlock (e.g. for every pod group, just start part of them), only the "last" 
pod group should be impacted.

For example, in google doc, the QueueJobController will create pods one by one based on QueueJob's
minAvaialbe; if minAvaialbe pods are not created, it'll NOT create pods for next QueueJob. If all minAvailable are meet,
QueueJobController will create pods up to `.spec.replicas` in round-robin for each QueueJob. 
There're still some cases are not covered, e.g. update minAvailabe for a better SLA, new high priority job,
start other jobs if can not meet minAvailable. That's another reason that prefer to focus on batch/run-to-complete firstly,
and introduce QueueJob; I'm also ok to enhance Job to support those cases in future, and maybe Queue.

Currently, AdmissionController checks pod based on the 'snapshot' of status/config, e.g. ResourceQuota. But for gang scheduling,
it also need to check whether controller **WILL** create more pods later to meet gang requirement. And a wrapper object can not
help as some AdmissionController are still doing pod level check.

Another corner case is that: a hug job (e.g. 1000+ tasks) waits for long running run-to-complete jobs (e.g. hours or days) :(
The candidate solution is to support Queue, and balance the resource sharing between Queue. This's also in backlog.